flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,247 @@
+
+  import KMeans._
+
+  /** Stores the learned clusters after the fit operation */
+  var centroids: Option[DataSet[LabeledVector]] = None
+
+  /**
+   * Sets the number of iterations.
+   *
+   * @param numIterations
+   * @return itself
+   */
+  def setNumIterations(numIterations: Int): KMeans = {
+    parameters.add(NumIterations, numIterations)
+    this
+  }
+
+  /**
+   * Sets the initial centroids on which the algorithm will start computing.
+   * These points should depend on the data and significantly influence the resulting centroids.
+   *
+   * @param initialCentroids A sequence of labeled vectors.
+   * @return itself
+   */
+  def setInitialCentroids(initialCentroids: DataSet[LabeledVector]): KMeans = {
+    parameters.add(InitialCentroids, initialCentroids)
+    this
+  }
+
+}
+
+/**
+ * Companion object of KMeans. Contains convenience functions, the parameter type definitions
+ * of the algorithm and the [[FitOperation]] & [[PredictOperation]].
+ */
+object KMeans {
+  val CENTROIDS = "centroids"
+
+  case object NumIterations extends Parameter[Int] {
+    val defaultValue = Some(10)
+  }
+
+  case object InitialCentroids extends Parameter[DataSet[LabeledVector]] {
+    val defaultValue = None
+  }
+
+  // ========================================== Factory methods ====================================
+
+  def apply(): KMeans = {
+    new KMeans()
+  }
+
+  // ========================================== Operations =========================================
+
+  /**
+   * [[PredictOperation]] for vector types. The result type is a [[LabeledVector]].
+   */
+  implicit def predictValues = {
+    new PredictOperation[KMeans, Vector, LabeledVector] {
+      override def predict(
+        instance: KMeans,
+        predictParameters: ParameterMap,
+        input: DataSet[Vector])
+      : DataSet[LabeledVector] = {
+
+        instance.centroids match {
+          case Some(centroids) => {
+            input.map(new SelectNearestCenterMapper).withBroadcastSet(centroids, CENTROIDS)
+          }
+
+          case None => {
+            throw new RuntimeException("The KMeans model has not been trained. Call first fit" +
+              "before calling the predict operation.")
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * [[FitOperation]] which iteratively computes centroids that match the given input DataSet by
+   * adjusting the given initial centroids.
+   */
+  implicit def fitKMeans = {
+    new FitOperation[KMeans, Vector] {
+      override def fit(
+        instance: KMeans,
+        fitParameters: ParameterMap,
+        input: DataSet[Vector])
+      : Unit = {
+        val resultingParameters = instance.parameters ++ fitParameters
+
+        val centroids: DataSet[LabeledVector] = resultingParameters.get(InitialCentroids).get
+        val numIterations: Int = resultingParameters.get(NumIterations).get
+
+        val finalCentroids = centroids.iterate(numIterations) { currentCentroids =>
```

Comment:
it would be nice is to have a convergence criterion as well, so that iterations stop early if the algorithm converges. Again you can check out the implementation GradientDescent on how to perform iterations with a convergence criterion.